onnxruntime/core/providers/cpu/cpu_execution_provider.cc

```diff
@@ -1784,6 +1792,7 @@ Status RegisterOnnxOperatorKernels(KernelRegistry& kernel_registry) {
     BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 13, float, Mean)>,
     BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 13, float, Gemm)>,
     BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 13, double, Gemm)>,
+    BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 13, MLFloat16, Gemm)>,
```

Comment:
>MLFloat16 [](http://example.com/codeflow?start=104&length=9)

new fp16 operators should be registered here. We should not rely on eigen implementation when there is no hardware support. We should let the graph optimizer choose the fp32 version to get higher performance

Reply:
In the new fp16 gemm implementation defined in gemm.cc, we let eigen handles very rare programming paths as a form of technical debt, e.g., strange shaped bias, which almost never happens.  We need to go back and reimplement those if in the future we do find those program paths important.

But it's not good to register a fp16 operator to the cpu EP when there is no hardware support. These are two different issues.

Reply:
#ifdef is processed at compile time. At then it has no knowledge of if the hardware supports fp16 or not. The macros is more about if the compiler can understand fp16.

Reply:
Updated.